### ISSUE TITLE
AnyUrl escapes percent sign in URLs without TLD that are already escaped.

### ISSUE THREAD

**Issue by alliefitter** [pydantic/pydantic#4458]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When using `AnyUrl` with a localhost URL (or any URL without a top level domain), if the provided URL already is encoded/escaped, pydantic will escape the percent sign again. When the domain has a TLD, it works perfectly fine. I've confirmed that the bug isn't present in 1.9.1. It seems to have been introduced in 1.10.0 and is still present in 1.10.1.

### Example Code

```Python
from pydantic import BaseModel, AnyUrl


class Foo(BaseModel):
    url: AnyUrl


class Bar(BaseModel):
    url: str


assert Foo(url='https://example.com/Foo%20Bar').url == 'https://example.com/Foo%20Bar'  # Passes
assert Bar(url='https://localhost/Foo%20Bar').url == 'https://localhost/Foo%20Bar'  # Passes
assert Foo(url='https://localhost/Foo%20Bar').url == 'https://localhost/Foo%20Bar'  # Fails: https://localhost/Foo%2520Bar
```

### Python, Pydantic & OS Version

```Text
$ python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.10.1
            pydantic compiled: True
                 install path: /home/afitter/.local/lib/python3.10/site-packages/pydantic
               python version: 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0]
                     platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.35
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [X] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Good catch, thank you for reporting, caused by #3061.

@FaresAhmedb what do you think?

I'm not on my laptop, so can't confirm until tomorrow.

**Comment by FaresAhmedb** [comment]
It seems that `AnyUrl.build` only gets called when there's no TLD.

```python
from pydantic import BaseModel, AnyUrl

class DebugAnyUrlBuild(AnyUrl):
    def build(**kwargs):
        print(f'{kwargs=}')
        return AnyUrl.build(**kwargs)

class Foo(BaseModel):
    url: DebugAnyUrlBuild

Foo(url='https://localhost/foo%20bar')
# > kwargs={'scheme': 'https', 'user': None, 'password': None, 'host': 'localhost', 'tld': None, 'host_type': 'int_domain', 'port': None, 'path': '/foo%20bar', 'query': None, 'fragment': None}

Foo(url='https://example.com/foo%20bar')
# >
```

While `AnyUrl.build` should't re-quote already encoded characters (_maybe?_, can be fixed by calling `urllib.parse.unquote` before quoting), I wonder why `AnyUrl.build` only gets called when there's no TLD.

**Comment by hramezani** [comment]
I can confirm the bug.

Yes, we `re-quote` already quoted strings.

another possible solution would be to check if the string is quoted and then prevent re-quoting.

**Comment by vetedde** [comment]
Here url is None when no TLD and than it calls build
```
class AnyUrl(str):
    ...

    @no_type_check
    def __new__(cls, url: Optional[str], **kwargs) -> object:
        return str.__new__(cls, cls.build(**kwargs) if url is None else url)
```

**Comment by alliefitter** [comment]
Maybe a solution could be something along the lines of...
```
if unquote(url) == url:
    url = quote(url)
```
Though, @FaresAhmedb's suggestion of just unquoting before quoting (`quote(unquote(url))`) would probably be more straight forward and cover more bases. I'm not aware of any side effects that would happen from doing that.

**Comment by samuelcolvin** [comment]
Thanks so much for looking into this.

I'm really sorry, it seems the originally implementation was not satisfactory, I made a mistake in merging it when it wasn't ready.

I've implemented #4469 which tries to solve the problem but also provides a bunch more test cases to ensure behaviour.

I've also noticed that we probably made the wrong mistake by going with `quote_plus=False` by default - e.g. the behaviour does not match chrome or httpie and I think does not match what is suggested by RFC 3986, again this is my fault. However I don't want to change this now, it can wait for V2.

Lastly the initial implementation wrongly quoted `/` in query parameters, that is fixed in #4469.

---

Please review and test #4469 and hopefully we can get this right (`quote_plus` issue apart) in a patch release.

**Comment by samuelcolvin** [comment]
After some consideration and discussions, I've decided to fix this by reverting #4224 in #4470.

See #4470 for a full explanation.
